### PR TITLE
pkg/storage: remove Exists from exported Backend interface

### PR DESCRIPTION
**Pull request by marwan-at-work**
Prerequisite to External Storage

Fixes https://github.com/gomods/athens/issues/1578

### CODE REVIEW

File: pkg/stash/with_redis.go [hunk 1]
```diff
@@ -57,13 +57,13 @@ func (s *redisLock) Stash(ctx context.Context, mod, ver string) (newVer string,
...
-	ok, err := s.checker.Exists(ctx, mod, ver)
-	if err != nil {
-		return ver, errors.E(op, err)
-	}
-	if ok {
+	_, err = s.getter.Info(ctx, mod, ver)
+	if err == nil {
 		return ver, nil
 	}
+	if errors.Is(err, errors.KindNotFound) {
```
Comment (marwan-at-work) [inline pkg/stash/with_redis.go]:
```suggestion
	if !errors.Is(err, errors.KindNotFound) {
```

File: pkg/stash/stasher.go [hunk 2]
```diff
@@ -58,13 +58,13 @@ func (s *stasher) Stash(ctx context.Context, mod, ver string) (string, error) {
...
-			return "", errors.E(op, err)
-		}
-		if exists {
+		_, err = s.storage.Info(ctx, mod, v.Semver)
+		if err == nil {
 			return v.Semver, nil
 		}
+		if !errors.Is(err, errors.KindNotFound) {
+			return "", errors.E(op, err)
+		}
```
Comment (arschles) [inline pkg/stash/stasher.go]: @marwan-at-work can you remind me why we never had to check for other errors?

File: pkg/stash/stasher_test.go [hunk 3]
```diff
@@ -90,9 +91,12 @@ func (ms *mockStorage) Save(ctx context.Context, module, version string, mod []b
 	return nil
 }
 
-func (ms *mockStorage) Exists(ctx context.Context, mod, ver string) (bool, error) {
+func (ms *mockStorage) Info(ctx context.Context, mod, ver string) ([]byte, error) {
 	ms.existsCalled = true
-	return ms.existsResponse, nil
+	if ms.existsResponse {
+		return []byte{'e', 'x', 'i', 's', 't', 's'}, nil
+	}
```
Comment (arschles) [inline pkg/stash/stasher_test.go]: can we change these to `infoCalled` and `infoResponse` respectively? seems important to change the names to indicate that `Info` was called to test existence

File: pkg/stash/stasher.go (hunk 2 above)
Comment (marwan-at-work) [reply pkg/stash/stasher.go]: @arschles there are no other errors, just not found and 500..and the not found is not really smart either thanks to `go mod download` not parsing VCS errors (which i understand is tricky)

**Comment by marwan-at-work** [comment]
@arschles 

>Also, sorry for the dumb question, but how is mockStorage a storage.Backend? I only see Save and Info defined on it

Because mockStorage embeds the storage.Backend interface, it doesn't need to declare all the methods, just the ones we need to use.

**Comment by marwan-at-work** [comment]
Closing the issue in favor of the alternative

**Pull request closed without merging**